File: playerC++/player.cpp
#include "libnetwork.hpp"
#include <string>
#include <iostream>
#include <cstdlib>
#include <ctime>



class Player{
private:
  Network net;
  
public:
  int nbLig,nbCol;
  char ** laby;
  int ** bombes;
  char ID;
  bool mortSubite;
  std::string ordres;
  std::string login;
  

  Player(const std::string& host, int port, const std::string& _login){
    ordres="SNEOB";
    net=Network();
    net.connectTo(host,port);
    login=_login;
    srand(time(NULL)); 
  }
  ~Player(){
    std::cout<<"Deconnexion du serveur"<<std::endl;
    net.disconnect();
  }

  void printSpecLaby(){
    std::cout<<"taille : "<<nbLig<<" x "<<nbCol<<std::endl;
    std::cout<<"ID : "<<ID<<std::endl;
  }

  void printLaby(){
      int i,j;
      std::cout<<"+";
      for (j=0; j<nbCol; j++) std::cout<<"-";
      std::cout<<"+"<<std::endl;

      for (i=0; i<nbLig; i++){
	std::cout<<"|";
	for (j=0; j<nbCol; j++)
	  std::cout<<laby[i][j];
	std::cout<<"|"<<std::endl;	
      }
      std::cout<<"+";
      for (j=0; j<nbCol; j++) std::cout<<"-";
      std::cout<<"+"<<std::endl;
  }
  
  void printBombes(){
    std::cout<<"liste des bombes : ";
    for (int i=0; i<nbLig; i++) 
      for (int j=0; j<nbCol; j++)
	if (bombes[i][j]==1)
	  std::cout<<"("<<i<<","<<j<<") ";
    std::cout<<std::endl;
  }

  void lireLaby(){
    for (int i=0; i<nbLig; i++) 
      for (int j=0; j<nbCol; j++) {
	laby[i][j]=net.getInt();
	bombes[i][j]=0;
      }
    int nbBombes=net.getInt();
    for (int i=0; i<nbBombes; i++){
      int l=net.getInt();
      int c=net.getInt();
      bombes[l][c]=1;
    }
  }

  void lireSpecLaby(){
    ID=net.getChar();
    nbLig=net.getInt();
    nbCol=net.getInt();
    std::string typePartie=net.getString();
    mortSubite= typePartie=="ms";
    laby=new char*[nbLig];
    bombes=new int*[nbLig];
    for (int i=0;i<nbLig;i++){
      laby[i]=new char[nbCol];
      bombes[i]=new int[nbCol];
    }
  }

  void connexionServeur(){
    net.sendString("LOGIN "+login);
    std::string reponse=net.getString();
    if (reponse=="SPEC"){
      std::cout<<"Connexion ok"<<std::endl;
      std::cout<<"Reception taille laby!"<<std::endl;
      lireSpecLaby();
      printSpecLaby();
      if (mortSubite)
	std::cout<<"partie mort subite"<<std::endl;
      else
	std::cout<<"partie standard"<<std::endl;
    }
    else{
       std::cout<<"Problème de connexion!!!"<<std::endl;
       std::cout<<"message reçu "<<reponse<<std::endl;
       exit(0);
    }
  }
  
  void demarrer(){
    bool encore=true;
    while(encore){
      std::string reponse=net.getString();
      std::cout<<"reçu : "<<reponse<<std::endl;
      if (reponse=="MAP"){
	std::cout<<"Reception map!"<<std::endl;
	lireLaby();
	printLaby();
	printBombes();
	// ici remplacer par le code de votre IA
	int r = rand()%5;
	std::cout<<"Ordre envoyé "<<ordres[r]<<std::endl;
	net.sendChar(ordres[r]);
      }
      else if (reponse=="DEAD" or reponse=="QUIT"){
	std::cout<<"Fin de la partie! serveur a dit "<<reponse<<std::endl;
	encore=false;
      }
      else
	std::cout<<"Message inattendu "<<reponse<<std::endl;
    }
    net.disconnect();
  }
};

int main(int argc, char *argv[]){
  if (argc!= 4){
    std::cerr<<"usage: %s HOST PORT LOGIN"<<std::endl;
    return -1;
  }
  int port=atoi(argv[2]);
  Player p=Player(std::string(argv[1]), port, std::string(argv[3]));
  p.connexionServeur();
  p.demarrer();
}

File: playerP/player.py
#! /usr/bin/python3
import time
import random
import sys
import os
from mySocket import *

ordres=['N','S','E','O','B']
laby=None
bombes=None
ID=-1
mortSubite=False
nbLig=-1
nbCol=-1

def printSpecLaby(nbLig,nbCol,ID):
    print('taille :',nbLig,'x',nbCol)
    print('ID :',ID)
            
def printLaby(laby):
    nbCol=len(laby[0])
    ligne='+'+'-'*nbCol+'+'
    print(ligne)
    for i in range(len(laby)):
        print('|',end='')
        for j in range(nbCol):
            print(laby[i][j],end='')
        print('|')
    print(ligne)

def printBombes(bombes):
    print("liste des bombes : ",end='')
    for i in range(len(bombes)):
        for j in range(len(bombes[0])):
            if bombes[i][j]==1:
                print("(",i,",",j,") ",sep='',end='')
    print()
                
def lireSpecLaby(net):
    ID=socket.getChar()  
    nbLig=socket.getInt()
    nbCol=socket.getInt()
    typePartie=socket.getString()
    mortSubite=typePartie=="ms"
    laby=[]
    bombes=[]
    for i in range(nbLig):
        laby.append([' ']*nbCol)
        bombes.append([0]*nbCol)
    return ID,nbLig,nbCol,mortSubite,laby,bombes


def lireLaby(net,nbLig,nbCol,laby,bombes):
    for i in range(nbLig):
        for j in range(nbCol):
            laby[i][j]=net.getChar()
            bombes[i][j]=0
    nbBombes=net.getInt()
    for i in range(nbBombes):
        lig=net.getInt()
        col=net.getInt()
        bombes[lig][col]=1
        
def connexionServeur(net,login):
    socket.send("LOGIN "+login)
    typeMsg=net.getString()
    print("Reçu :",typeMsg)
    if typeMsg=="SPEC": 
        print('connexion réussie')
        print('Reception taille laby!')
        ID,nbLig,nbCol,mortSubite,laby,bombes=lireSpecLaby(net)
        printSpecLaby(nbLig,nbCol,ID)
        if mortSubite:
            print("partie mort subite")
        else:
            print("partie standard")
    else:
        print("Problème de connexion!!!")
        print("message reçu ",typeMsg)
        net.close()
        sys.exit(1)
    return ID,nbLig,nbCol,mortSubite,laby,bombes
    


def demarrer(net,ID,nbLig,nbCol,mortSubite,laby,bombes):
    fini=False
    while(not fini):
        typeMsg=net.getString()
        print("Reçu :",typeMsg)
        if typeMsg=='MAP':
            lireLaby(net,nbLig,nbCol,laby,bombes)
            printLaby(laby)
            printBombes(bombes)
            # ici remplacer par le code de votre IA
            ordre=random.choice(ordres)
            print('Ordre envoyé',ordre)
            net.send(ordre)
        elif typeMsg in ['QUIT','DEAD']:
            print("fin de la partie",typeMsg)
            fini=True
        else:
            print('message du serveur inattendu',typeMsg)
    net.close()
        
        
opts=sys.argv[1:]
socket=MySocket()
socket.connect(opts[0],int(opts[1]))
ID,nbLig,nbCol,mortSubite,laby,bombes=connexionServeur(socket,opts[2])
demarrer(socket,ID,nbLig,nbCol,mortSubite,laby,bombes)

